File: lib/features/transactions/domain/models/income.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:savvy/core/constants/financial_enums.dart';

part 'income.freezed.dart';
part 'income.g.dart';

@freezed
abstract class Income with _$Income {
  const factory Income({
    required String id,
    required double amount,
    required IncomeCategory category,
    String? person,
    String? source,
    required DateTime date,
    String? note,
    @Default(false) bool isRecurring,
    DateTime? recurringEndDate,
    /// true ise [amount] brüt tutardır; net, ay bazında hesaplanır.
    @Default(false) bool isGross,
    @Default(false) bool isDeleted,
    required DateTime createdAt,
    /// Per-month amount overrides for recurring items.
    /// Key: "YYYY-MM", Value: override amount for that month.
    /// Months not present use the default [amount].
    @Default({}) Map<String, double> monthlyOverrides,
    /// true = alındı (gelir tahsil edildi), false = beklemede.
    /// Tek seferlik işlemler için kullanılır.
    @Default(false) bool isSettled,
    /// Recurring işlemlerin ay bazlı settled durumu.
    /// Key: "YYYY-MM", Value: true = alındı.
    /// Map'te olmayan aylar isSettled default'unu kullanır.
    @Default({}) Map<String, bool> settledMonths,
  }) = _Income;

  factory Income.fromJson(Map<String, dynamic> json) => _$IncomeFromJson(json);
}

File: SavvyApp/Packages/Core/SavvyFoundation/Sources/SavvyFoundation/Models/Income.swift
import Foundation

public struct Income: FirestoreEntity {
    public let id: String
    public var amount: Decimal
    public var category: IncomeCategory
    public var person: String?
    public var source: String?
    public var date: Date
    public var note: String?
    public var isRecurring: Bool
    public var recurringEndDate: Date?
    public var monthlyOverrides: [String: Decimal]
    public var isGross: Bool
    public var isDeleted: Bool
    public let createdAt: Date

    public init(
        id: String = UUID().uuidString,
        amount: Decimal,
        category: IncomeCategory,
        person: String? = nil,
        source: String? = nil,
        date: Date = Date(),
        note: String? = nil,
        isRecurring: Bool = false,
        recurringEndDate: Date? = nil,
        monthlyOverrides: [String: Decimal] = [:],
        isGross: Bool = false,
        isDeleted: Bool = false,
        createdAt: Date = Date()
    ) {
        self.id = id
        self.amount = amount
        self.category = category
        self.person = person
        self.source = source
        self.date = date
        self.note = note
        self.isRecurring = isRecurring
        self.recurringEndDate = recurringEndDate
        self.monthlyOverrides = monthlyOverrides
        self.isGross = isGross
        self.isDeleted = isDeleted
        self.createdAt = createdAt
    }
}
